### ISSUE TITLE
[1.13] Deadlock in SetClusterProvider/libnetwork callbacks

### ISSUE THREAD

**Issue by tonistiigi** [docker/docker#29952]
https://jenkins.dockerproject.org/job/Docker-PRs-experimental/29248/console (from #29926) shows a failed test that hit a deadlock.

Daemon logs with traces: https://gist.github.com/tonistiigi/2ae3675901f18c3ccab7e6378aeaaf74

The issue seems to be that we are calling `SetClusterProvider()` while holding the cluster lock, the implementation needs a lock for a controller. Now the networking stack may call back to cluster `IsManager()` while holding the controller lock. `IsManager()` requires readlock for cluster.

This should be one of the cases solved with https://github.com/docker/docker/pull/28907 but that PR was not picked for `v1.13`.

We should evaluate the priority of this and see if there is a easy workaround that we could include in `v1.13`. One of the options could be to call `SetClusterProvider()` in a goroutine if it only works as a event dispatcher(cc @mavenugo to see if it would be safe).

There is also a possibility that recent changes in swarmkit (around graceful shutdown for example) have made this race more likely to appear.

@aaronlehmann @mavenugo @dongluochen

**Comment by aaronlehmann** [comment]
IMHO we should treat this as a 1.13 blocker unless we have data to show that it rarely triggers.

**Comment by aboch** [comment]
One way to manually reproduce the deadlock is to concurrently run 
`docker swarm leave --force` and `docker swarm init -force-new-cluster` 
few times back to back on  a manager.

I could confirm that with the fix suggested by @tonistiigi (call `SetClusterProvider()` as a go routine) and the panic fix by @aaronlehmann (docker/swarmkit/pull/1848)  I can no longer reproduce the deadlock.

Will need some more testing, but I think the fix is ok for 1.13.

**Comment by tonistiigi** [comment]
I managed to get the following with `docker swarm leave --force & docker swarm init & docker swarm init --force-new-cluster`. Goroutine would not help against that but not sure if it is also possible on a practical case.

```
goroutine 292 [semacquire]:
sync.runtime_Semacquire(0xc42045e868)
        /usr/local/go/src/runtime/sema.go:47 +0x30
sync.(*Mutex).Lock(0xc42045e864)
        /usr/local/go/src/sync/mutex.go:85 +0xd0
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).getStores(0xc42045e780, 0x0, 0x0,
 0x0)
        /go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/store.go:75 +0x58
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).getNetworksFromStore(0xc42045e780
, 0xc4204d7280, 0xc4209aec78, 0x6391cf, 0x2536060, 0x2a2d5364)
        /go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/store.go:144 +0x61
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).Networks(0xc42045e780, 0x0, 0x1f,
 0x5dc)
        /go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/controller.go:838 +0x4b
github.com/docker/docker/daemon.(*Daemon).V4Subnets(0xc4200c6200, 0x5, 0x8, 0x0)
        /go/src/github.com/docker/docker/daemon/daemon.go:869 +0x4f
github.com/docker/docker/daemon/cluster.(*Cluster).resolveSystemAddrViaSubnetCheck(0xc4201003c0, 0x197f6a0,
0x4, 0x6, 0x0, 0x2)
        /go/src/github.com/docker/docker/daemon/cluster/listen_addr.go:165 +0x9b
github.com/docker/docker/daemon/cluster.(*Cluster).resolveSystemAddr(0xc4201003c0, 0x10, 0x10, 0x24f52a0, 0x
c4200108a0, 0x0)
        /go/src/github.com/docker/docker/daemon/cluster/listen_addr_linux.go:85 +0x955
github.com/docker/docker/daemon/cluster.(*Cluster).resolveAdvertiseAddr(0xc4201003c0, 0x0, 0x0, 0xc42102666e
, 0x4, 0x4, 0x0, 0x0, 0x0, 0xc4210204f8, ...)
        /go/src/github.com/docker/docker/daemon/cluster/listen_addr.go:99 +0x2d3
github.com/docker/docker/daemon/cluster.(*Cluster).Init(0xc4201003c0, 0xc421026666, 0xc, 0x0, 0x0, 0x1, 0x1983846, 0x7, 0x0, 0x0, ...)
```

**Comment by aboch** [comment]
@tonistiigi On a linux platform, this should happen only when docker is run on a system with no `device` interface, like when it is run inside a container.

**Comment by aboch** [comment]
Fixed by #30038
